Support new flag: hideExactTimes

pnuts2 opened the pull request:
This closes https://redmine.buzzhives.com/issues/16205

nighthawk commented on Sources/TripKit/model/API/RoutingAPIModel.swift:
```diff
@@ -35,6 +35,7 @@ extension TKAPI {
   public struct Trip: Codable, Hashable {
     @ISO8601OrSecondsSince1970 public var depart: Date
     @ISO8601OrSecondsSince1970 public var arrive: Date
+    public let hideExactTimes: Bool?
```
Could just be a `Bool` and we default it to `false`. That gives us just two states to deal with:

```suggestion
    @DefaultFalse public var hideExactTimes: Bool
```

nighthawk commented on Sources/TripKit/server/parsing/TKRoutingParser.swift:
```diff
@@ -248,6 +248,7 @@ public final class TKRoutingParser {
         
         trip.departureTime = apiTrip.depart
         trip.arrivalTime = apiTrip.arrive
+        trip.hideExactTimes = apiTrip.hideExactTimes ?? false
```
```suggestion
        trip.hideExactTimes = apiTrip.hideExactTimes
```

nighthawk reviewed (changes requested):
Also need to hide times on:

- Trip overview map
- Mode-by-mode header

nighthawk commented:
> From the ticket, it seems the `hideExactTimes` only applies to routing results and trip overview, which has already been done.

Yeah, the ticket doesn't mention them, but it's about removing all explicit times, so we should remove them all. Map was an oversight, that it wasn't listed.

For the MxM: It should be replaced with the leave after or arrive by time; whichever was used to plan the trip.

pnuts2 commented:
> > From the ticket, it seems the `hideExactTimes` only applies to routing results and trip overview, which has already been done.
> 
> Yeah, the ticket doesn't mention them, but it's about removing all explicit times, so we should remove them all. Map was an oversight, that it wasn't listed.

![Simulator Screen Shot - iPhone 13 - 2021-12-06 at 21 11 27](https://user-images.githubusercontent.com/2722800/144827986-7504f59e-9d4d-42b0-9ac2-fb0fc77a850a.png)

So in the above screen, what changes need to be made? Remove 10:00pm or remove the text "for 4 min"? Can you point me to the data type that produces the 10:00pm overlay?
 
> For the MxM: It should be replaced with the leave after or arrive by time; whichever was used to plan the trip.

pnuts2 commented on Sources/TripKitUI/views/map annotations/TKUIAnnotations+TripKit.swift:
```diff
@@ -153,7 +153,7 @@ extension TKSegment: TKUISemaphoreDisplayable {
       }
     } else {
-      return .headWithTime(departureTime, timeZone, isRealTime: timesAreRealTime)
+      return trip.hideExactTimes ? .headOnly : .headWithTime(departureTime, timeZone, isRealTime: timesAreRealTime)
```
![Simulator Screen Shot - iPhone 13 - 2021-12-06 at 21 34 57](https://user-images.githubusercontent.com/2722800/144831437-26408bb6-c3a5-4236-a437-716c33529e2e.png)

@nighthawk Can we do this here, assuming hide the text on the `timeLabel` is what is required here.

nighthawk replied:
Yep, looks good.

nighthawk commented on Sources/TripKitUI/helper/Trip+Titles.swift:
```diff
@@ -23,16 +23,19 @@ extension Trip {
     let duration = arrival.durationSince(departure)
     
     if focusOnDuration {
+      guard !hideExactTimes else { return (duration, "") }
```
This should be done outside the `if`, otherwise a combination of `focusOnDuration == false && hideExactTimes` would still show exact times.

pnuts2 replied:
Oh okay. I can make this change, though I thought we'd only want to hide this when we want to focus on duration, e.g., in those cases where there's no fixed departure and arrival like PT or frequency-based services?

pnuts2 commented:
@nighthawk Addressed your comment above, about this requirement

> For the MxM: It should be replaced with the leave after or arrive by time; whichever was used to plan the trip.

according to Mariano's update [here](https://redmine.buzzhives.com/issues/16205#note-8), I'll need to add a few more fields to a `Trip`. Should we use another PR for that, so this one focuses on this API change?

> adding `hideExactTimes`

nighthawk commented:
> @nighthawk Addressed your comment above, about this requirement
> 
> > For the MxM: It should be replaced with the leave after or arrive by time; whichever was used to plan the trip.
> 
> according to Mariano's update [here](https://redmine.buzzhives.com/issues/16205#note-8), I'll need to add a few more fields to a `Trip`. Should we use another PR for that, so this one focuses on this API change only?

TripKit already has this information. It's in `trip.request` and automatically set when planning a trip. What Mariano meant to also add this to `trip/update` output, but TripKit is ready already for this, too, by using `TKAPI.Query` and `TKRoutingParser+Populate.swift`.

pnuts2 commented:
@nighthawk This is finally done. In this PR

* Exact times are hidden when appropriate.
* Pick up window is displayed
* Pick up and drop off cell is displayed as per Menlo

Please do another round of review. Cheers

nighthawk reviewed (approved):
LGTM
